qiskit/transpiler/passes/optimization/commutation_analysis.py

```diff
@@ -115,8 +116,16 @@ def _commute(node1, node2):
 
     id_op = Operator(np.eye(2 ** qbit_num))
 
-    op12 = id_op.compose(node1.op, qargs=qarg1).compose(node2.op, qargs=qarg2)
-    op21 = id_op.compose(node2.op, qargs=qarg2).compose(node1.op, qargs=qarg1)
+    if ((id(node1.op), str(qarg1)), (id(node2.op), str(qarg2))) in cache:
+        op12 = cache[((id(node1.op), str(qarg1)), (id(node2.op), str(qarg2)))]
+    else:
+        op12 = id_op.compose(node1.op, qargs=qarg1).compose(node2.op, qargs=qarg2)
+        cache[((id(node1.op), str(qarg1)), (id(node2.op), str(qarg2)))] = op12
+    if ((id(node2.op), str(qarg2)), (id(node1.op), str(qarg1))) in cache:
+        op21 = cache[((id(node2.op), str(qarg2)), (id(node1.op), str(qarg1)))]
+    else:
+        op21 = id_op.compose(node2.op, qargs=qarg2).compose(node1.op, qargs=qarg1)
+        cache[((id(node2.op), str(qarg2)), (id(node1.op), str(qarg1)))] = op21
```

Comment:
This is a bit tricky to parse at the moment. Something like
```
cache_key = (...)
if cache_key not in cache:
  cache[cache_key] = id_op.compose(...)
op = cache[cache_key]
```
might be easier to read, at the cost of an additional dict lookup when populating the cache (but you'd save one cost of building the cache key when hitting the cache).

Reply:
I did this to a certain extent in https://github.com/Qiskit/qiskit-terra/pull/3878/commits/380b4e9ea34febeed036c925aa0c15899ba8f1af